core/src/main/java/com/linecorp/armeria/internal/common/stream/ByteBufsInputStream.java

```diff
@@ -150,8 +150,13 @@ public void setEos() {
     public void add(ByteBuf buffer) {
         requireNonNull(buffer, "buffer");
         if (eos.get()) {
+            buffer.release();
             throw new IllegalStateException("Already closed");
         }
+        if (!buffer.isReadable()) {
+            buffer.release();
+            return;
+        }
         if (!buffers.add(buffer)) {
             throw new IllegalStateException("Unable to add new buffer");
```

Comment:
The `buffer` should also be released before raising `IllegalStateException("Unable to add new buffer")`.